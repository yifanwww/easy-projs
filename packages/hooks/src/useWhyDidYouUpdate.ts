import { useEffect, useRef } from 'react';

export function useWhyDidYouUpdate<P extends Record<string, unknown>>(name: string, props: P): void {
    const prevPropsRef = useRef<P>();

    useEffect(() => {
        const prevProps = prevPropsRef.current;

        if (prevProps === undefined) {
            console.log(`[why-did-you-update] '${name}' first render.`);
        } else {
            const keys = Object.keys({ ...prevProps, ...props });
            const changes: Record<string, { from: unknown; to: unknown }> = {};

            for (const key of keys) {
                if (prevProps[key] !== props[key]) {
                    changes[key] = { from: prevProps[key], to: props[key] };
                }
            }

            if (Object.keys(changes).length) {
                console.log(`[why-did-you-update] '${name}'`, changes);
            }
        }

        prevPropsRef.current = props;
    });
}
