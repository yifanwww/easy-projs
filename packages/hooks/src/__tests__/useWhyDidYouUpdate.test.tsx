import { render } from '@testing-library/react';

import { useWhyDidYouUpdate } from '../useWhyDidYouUpdate';

beforeEach(() => {
    jest.spyOn(console, 'log').mockImplementation((...params) => expect(params).toMatchSnapshot());
});

describe(`Test react hook \`${useWhyDidYouUpdate.name}\``, () => {
    it('works with simple props', () => {
        interface IProps {
            fontSize: number | string;
            title: string;
        }

        function TestComponent(props: IProps) {
            useWhyDidYouUpdate(TestComponent.name, props);
            return <div />;
        }

        const { rerender } = render(<TestComponent fontSize="20px" title="title" />);
        rerender(<TestComponent fontSize={20} title="title" />);
        rerender(<TestComponent fontSize={20} title="new-title" />);
        rerender(<TestComponent fontSize="28px" title="new-title" />);
    });

    it('works with complex props', () => {
        interface IProps {
            color: string;
            style?: React.CSSProperties;
            title: string;
        }

        function TestComponent(props: IProps) {
            useWhyDidYouUpdate(TestComponent.name, props);
            return <div />;
        }

        const { rerender } = render(<TestComponent color="red" title="title" />);
        rerender(<TestComponent color="red" title="title" style={{ fontSize: '28px' }} />);
        rerender(<TestComponent color="red" title="new-title" style={{ fontSize: '28px', margin: '8px 0px' }} />);
        rerender(<TestComponent color="green" title="new-title" style={{ fontSize: '28px', margin: '8px 0px' }} />);
    });
});
