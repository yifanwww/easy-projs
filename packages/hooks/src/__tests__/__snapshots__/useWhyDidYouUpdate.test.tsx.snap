// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test react hook \`useWhyDidYouUpdate\` works with complex props 1`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent' first render.",
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with complex props 2`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent'",
    Object {
      "style": Object {
        "from": undefined,
        "to": Object {
          "fontSize": "28px",
        },
      },
    },
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with complex props 3`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent'",
    Object {
      "style": Object {
        "from": Object {
          "fontSize": "28px",
        },
        "to": Object {
          "fontSize": "28px",
          "margin": "8px 0px",
        },
      },
      "title": Object {
        "from": "title",
        "to": "new-title",
      },
    },
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with complex props 4`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent'",
    Object {
      "color": Object {
        "from": "red",
        "to": "green",
      },
      "style": Object {
        "from": Object {
          "fontSize": "28px",
          "margin": "8px 0px",
        },
        "to": Object {
          "fontSize": "28px",
          "margin": "8px 0px",
        },
      },
    },
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with simple props 1`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent' first render.",
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with simple props 2`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent'",
    Object {
      "fontSize": Object {
        "from": "20px",
        "to": 20,
      },
    },
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with simple props 3`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent'",
    Object {
      "title": Object {
        "from": "title",
        "to": "new-title",
      },
    },
  ],
}
`;

exports[`Test react hook \`useWhyDidYouUpdate\` works with simple props 4`] = `
Object {
  "console": "log",
  "messages": Array [
    "[why-did-you-update] 'TestComponent'",
    Object {
      "fontSize": Object {
        "from": 20,
        "to": "28px",
      },
    },
  ],
}
`;
